lib/components/joystick/joystick_directional.dart

```diff
@@ -0,0 +1,290 @@
+      final double nextY = dist * sin(_radAngle);
+      final Offset nextPoint = Offset(nextX, nextY);
+
+      final Offset diff = Offset(_backgroundRect.center.dx + nextPoint.dx,
+              _backgroundRect.center.dy + nextPoint.dy) -
+          _knobRect.center;
+      _knobRect = _knobRect.shift(diff);
+
+      final double _intensity = dist / (_tileSize * _backgroundAspectRatio / 3);
+
+      if (_intensity == 0) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.IDLE,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+        return;
+      }
+
+      if (degrees > -22.5 && degrees <= 22.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_RIGHT,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if (degrees > 22.5 && degrees <= 67.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_DOWN_RIGHT,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if (degrees > 67.5 && degrees <= 112.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_DOWN,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if (degrees > 112.5 && degrees <= 157.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_DOWN_LEFT,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if ((degrees > 157.5 && degrees <= 180) ||
+          (degrees >= -180 && degrees <= -157.5)) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_LEFT,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if (degrees > -157.5 && degrees <= -112.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_UP_LEFT,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if (degrees > -112.5 && degrees <= -67.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_UP,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
+
+      if (degrees > -67.5 && degrees <= -22.5) {
+        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
+          directional: JoystickMoveDirectional.MOVE_UP_RIGHT,
+          intensity: _intensity,
+          radAngle: _radAngle,
+        ));
+      }
```

Comment:
a few things:
1. the first if has a return. the others don't. regardless nothing is done after all ifs and only one can match? so how about an if/else if/else if/else chain? or even a switch
2. it seems the only thing that changes is the JoystickMoveDirectional. the rest is the same. so I suggest a calculateJoystickMoveDirectional function that returns the JoystickMoveDirectional (does the switch/case or if/else chain). and then you can have the bulk of the code:
```
        _joystickController.joystickChangeDirectional(JoystickDirectionalEvent(
          directional: calculate....,
          intensity: _intensity,
          radAngle: _radAngle,
        ));
```
only once.